pkg/providers/k8s/k8s.go

```diff
@@ -0,0 +1,147 @@
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/helmfile/vals/pkg/api"
+	"github.com/helmfile/vals/pkg/log"
+)
+
+type provider struct {
+	log            *log.Logger
+	KubeConfigPath string
+	KubeContext    string
+}
+
+func New(l *log.Logger, cfg api.StaticConfig) *provider {
+	p := &provider{
+		log: l,
+	}
+
+	kubeConfig, err := getKubeConfig(cfg)
+	if err != nil {
+		fmt.Printf("An error occurred getting the Kubeconfig path: %s\n", err)
+		return p
+	}
+
+	p.KubeConfigPath = kubeConfig
+	p.KubeContext = getKubeContext(cfg)
+
+	return p
+}
+
+func getKubeConfig(cfg api.StaticConfig) (string, error) {
+	// Use kubeConfigPath from URI parameters if specified
+	if cfg.String("kubeConfigPath") != "" {
+		if _, err := os.Stat(cfg.String("kubeConfigPath")); err != nil {
+			return cfg.String("kubeConfigPath"), fmt.Errorf("kubeConfigPath URI parameter is set but path %s does not exist.", cfg.String("kubeConfigPath"))
+		}
+	}
+
+	// Use path in KUBECONFIG environment variable if set
+	if envPath := os.Getenv("KUBECONFIG"); envPath != "" {
+		if _, err := os.Stat(envPath); err != nil {
+			return envPath, fmt.Errorf("KUBECONFIG environment variable is set but path %s does not exist.", envPath)
+		}
+	}
+
+	// Use default kubeconfig path if it exists
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return "", fmt.Errorf("An error occurred getting the user's home directory: %s", err)
+	}
+
+	defaultPath := homeDir + "/.kube/config"
+	if _, err := os.Stat(defaultPath); err == nil {
+		return defaultPath, nil
+	}
+
+	return "", fmt.Errorf("No path was found in any of the following: kubeContext URI param, KUBECONFIG environment variable, or default path %s does not exist.", defaultPath)
+}
+
+func (p *provider) GetString(path string) (string, error) {
+	separator := "/"
+	splits := strings.Split(path, separator)
+
+	if len(splits) != 3 {
+		return "", fmt.Errorf("Invalid path %s. Path must be in the format <namespace>/<secret>/<key>", path)
+	}
+
+	namespace := splits[0]
+	secretName := splits[1]
+	key := splits[2]
+
+	if p.KubeConfigPath == "" {
+		return "", fmt.Errorf("No Kubeconfig path was found")
+	}
+
+	secretData, err := getSecret(namespace, secretName, p.KubeConfigPath, p.KubeContext, context.Background())
+	secret, exists := secretData[key]
+	if err != nil || !exists {
```

Comment:
hmmm... we should handle it separately. it will be better.

Reply:
Thanks. Separated this.